include/net/nrf_cloud_pgps.h

```diff
@@ -232,21 +235,63 @@ int nrf_cloud_pgps_find_prediction(struct nrf_cloud_pgps_prediction **prediction
 int nrf_cloud_pgps_request(const struct gps_pgps_request *request);
 
 /**@brief Requests all available P-GPS data from nRF Cloud via MQTT.
+ * Not used for other transports.
  *
  * @return 0 if successful, otherwise a (negative) error code.
  */
 int nrf_cloud_pgps_request_all(void);
-#endif /* CONFIG_NRF_CLOUD_MQTT */
 
-#if defined(CONFIG_NRF_CLOUD_PGPS_TRANSPORT_NONE)
+/**@brief Use this function when you use a custom transport mechanism for downloading prediction
+ * data. When your pgps_event_handler_t receives a PGPS_EVT_REQUEST event, send
+ * the P-GPS request to the cloud over your transport, then call this function.
+ * This function resets the internal state of the library and prepares the flash storage to receive
+ * incoming data.
+ *
+ * The library calls this automatically for HTTP(S) transports.
+ *
+ * @retval 0 Ready for updating.
+ * @retval -ENOMEM No free space in P-GPS flash storage area.
+ * @retval -EBUSY P-GPS download already in progress.
+ * @retval -ENODEV Flash device is not ready.
+ * @retval -EFAULT Error accessing flash stream.
+ * @return a negative value indicates an error.
+ */
+int nrf_cloud_pgps_begin_update(void);
+
+/**@brief Call this function when your transport receives a block of prediction
+ * data. Prediction data must be received in file order.
+ *
+ * The library calls this automatically for HTTP(S) transports.
+ *
+ * @param buf Pointer to P-GPS data received from transport.
+ * @param len Size of P-GPS data chunk received from transport.
+ * @retval 0 Buffer processed.
+ * @retval -EINVAL Invalid parameter to function or invalid header in P-GPS data received.
+ * @retval -ENODATA Unable to determine current date and time.
+ * @retval -ENOMEM No free flash in P-GPS storage area.
+ * @retval -EFAULT Error writing to flash stream.
+ * @return a negative value indicates an error.
+ */
+int nrf_cloud_pgps_process_update(uint8_t *buf, size_t len);
+
+/**@brief Call this function regardless of whether the download succeeded or failed.
+ *
+ * Not needed for custom download transports.
+ *
+ * @retval 0 No failure.
+ * @return a negative value indicates an error.
+ */
+int nrf_cloud_pgps_finish_update(void);
+
 /**@brief If previous request for P-GPS data failed, re-enable future retries.
  * This is should be called by the application after it attempts to
  * handle PGPS_EVT_REQUEST, but is unable to complete it successfully. For
  * example, it should be called if the cloud connection being used to transmit
  * the request is temporarily unavailable.
+ *
+ * No need to call when using the MQTT transport.
  */
 void nrf_cloud_pgps_request_reset(void);
```

Comment:
Since the Location lib supports external downloading of PGPS data (CONFIG_LOCATION_METHOD_GNSS_PGPS_EXTERNAL), I wonder if we should add wrappers to Location lib for calling these functions as well. Seems kind of stupid, but it might be architecturally more correct? Did you consider if these changes affect CONFIG_LOCATION_METHOD_GNSS_PGPS_EXTERNAL in some other way @plskeggs ?

Reply:
@hiltunent I did not even know about `CONFIG_LOCATION_METHOD_GNSS_PGPS_EXTERNAL`.  Thanks for pointing it out.

Reply:
OK.  This feature of the location library is only about the first stage of getting P-GPS data.  The first stage is that a request is sent to the cloud which indicates prediction count, prediction period, gps day, and gps time of day desired.  The response from the cloud is a small text document (JSON-like) that specifies the host name and path to the binary data file that contains the predictions in binary format.

The second stage is that the P-GPS library downloads the binary data from that host name and path, and stores it into flash.

This PR only affects the second stage -- the binary download, not the request/response portion.

The location library `CONFIG_LOCATION_METHOD_GNSS_PGPS_EXTERNAL` only manages the first stage, so the answer is no, you do not need to call these functions.